d2/src/main/java/com/linkedin/d2/balancer/dualread/UriPropertiesDualReadMonitor.java

```diff
@@ -0,0 +1,177 @@
+import java.util.function.Function;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class UriPropertiesDualReadMonitor {
+  private static final Logger LOG = LoggerFactory.getLogger(UriPropertiesDualReadMonitor.class);
+
+  private final Map<String, ClusterMatchRecord> _clusters = new HashMap<>();
+  // Limit error report logging to every 10 minutes
+  private final RateLimitedLogger RATE_LIMITED_LOGGER =
+      new RateLimitedLogger(LOG, TimeUnit.MINUTES.toMillis(10), SystemClock.instance());
+  private int _totalUris = 0;
+  private int _matchedUris = 0;
+  private final DualReadLoadBalancerJmx _dualReadLoadBalancerJmx;
+
+  public UriPropertiesDualReadMonitor(DualReadLoadBalancerJmx dualReadLoadBalancerJmx) {
+    _dualReadLoadBalancerJmx = dualReadLoadBalancerJmx;
+  }
+
+  public void reportData(String clusterName, UriProperties property, boolean fromNewLb) {
+    ClusterMatchRecord cluster = _clusters.computeIfAbsent(clusterName, k -> new ClusterMatchRecord());
+
+    if (fromNewLb) {
+      cluster._newLb = property;
+    } else {
+      cluster._oldLb = property;
+    }
+
+    _totalUris -= cluster._uris;
+    _matchedUris -= cluster._matched;
+
+    LOG.debug("Updated URI properties for cluster {}:\nOld LB: {}\nNew LB: {}",
+        clusterName, cluster._oldLb, cluster._newLb);
+
+    if (cluster._oldLb == null && cluster._newLb == null) {
+      _clusters.remove(clusterName);
+      updateJmxMetrics(clusterName, null);
+      return;
+    }
+
+    cluster._matched = 0;
+
+    if (cluster._oldLb == null || cluster._newLb == null) {
+      LOG.debug("Added new URI properties for {} for {} LB.", clusterName, fromNewLb ? "New" : "Old");
+
+      cluster._uris = (cluster._oldLb == null) ? cluster._newLb.Uris().size() : cluster._oldLb.Uris().size();
+      _totalUris += cluster._uris;
+
+      updateJmxMetrics(clusterName, cluster);
+      return;
+    }
+
+    cluster._uris = cluster._oldLb.Uris().size();
+    Set<URI> newLbUris = new HashSet<>(cluster._newLb.Uris());
+
+    for (URI uri : cluster._oldLb.Uris()) {
+      if (!newLbUris.remove(uri)) {
+        continue;
+      }
+
+      if (compareURI(uri, cluster._oldLb, cluster._newLb)) {
+        cluster._matched++;
+      }
+    }
+    // add the remaining unmatched URIs in newLbUris to the uri count
+    cluster._uris += newLbUris.size();
+
+    if (cluster._matched != cluster._uris) {
+      RATE_LIMITED_LOGGER.info("Mismatched cluster properties for {} (match score: {}, total uris: {}):"
+              + "\nOld LB: {}\nNew LB: {}",
+          clusterName, cluster._matched / cluster._uris, cluster._uris, cluster._oldLb, cluster._newLb);
+    }
+
+    _totalUris += cluster._uris;
+    _matchedUris += cluster._matched;
+
+    updateJmxMetrics(clusterName, cluster);
+  }
+
+  private void updateJmxMetrics(String clusterName, ClusterMatchRecord cluster) {
+    _dualReadLoadBalancerJmx.setClusterMatchRecord(clusterName, cluster);
```

Comment:
`ClusterMatchRecord` isn't thread safe. It's being modified by `reportData` in one thread, and read by the caller of `getClusterMatchRecord` in another thread. I would recommend something like this near the top of `reportData`:
```java
    ClusterMatchRecord cluster = _clusters.compute(clusterName, (k, v) -> {
      if (v == null) {
        return new ClusterMatchRecord();
      } else {
        return v.clone();
      }
    });
```
With the corresponding `clone` method added to `ClusterMatchRecord` of course.

Reply:
yeah, or we could simply call _dualReadLoadBalancerJmx.setClusterMatchRecord(clusterName, **cluster.clone()**) ?

Reply:
added the lock